Initialize Metabase with config from a file

camsaul opened the pull request:
Implements #2052

Metabase now supports initialization on launch from a config file named `config.yml` in the current directory (or `MB_CONFIG_FILE_PATH` if specified). This PR adds support for creating Users and Databases on launch. Here's an example file that I have working locally:

```yaml
version: 1
config:
  users:
    - first_name: Cam
      last_name: Era
      password: 2cans3cans4cans
      email: cam@example.com
      is_superuser: true
  databases:
    - name: test-data (Postgres)
      engine: postgres
      details:
        host: localhost
        port: 5432
        password: {{ env POSTGRES_TEST_DATA_PASSWORD }}
        db: test-data
    - name: Sample Dataset (Copy)
      engine: h2
      details:
        db: "/home/cam/metabase/resources/sample-database.db;USER=GUEST;PASSWORD={{env SAMPLE_DATASET_PASSWORD}}"
```

If the Users or Databases in the config file already exist, this will upsert the specified values, e.g. if a user with email `cam@example.com` already exists, it will `UPDATE` that User so its values match those specified in the config file. I'm not sure whether this makes more or less sense than just no-opping in that case. If I change my name from `Cam Era` to `Cam Saul` in Metabase, should restarting stomp on that change because the config file specifies `Cam Era`? It seems like stomping is bad, but it also seems like ignoring changes in a  config file, especially to stuff like DB details is bad as well.

You can pretty much set any column of the User or Database tables in the config file, including other stuff like the sync schedules for a Database; they're just taken from the YAML and `INSERT`ed directly into the application database. 

Env vars can be specified with `{{ template-tags }}` like `{{ env POSTGRES_TEST_DATA_PASSWORD }}` -- these reuse the code that parses template tags in SQL queries💪-- which means `[[options {{template-tags}}]]` are supported as well. In the future we can support more template types as well (such as splicing in the contents of other files, or other types of secrets, neat functions, etc.), but `env` seems like enough for v1. Recursive expansion is disallowed for now, so no env vars that have values like `{{ env ANOTHER_ENV_VAR }}`. I tried to take care to make sure the values of the env vars are not logged in case of an error, at least not by the new `metabase.config.file` code.

#### TODO before merge

There are still a few final things to do but I think someone can start reviewing it at this point in time.

- [x] tests for Database creation/upsert
- [ ] Feature flag behind appropriate Enterprise/Pro token feature (TBD)

#### Open questions

- [ ] Should launching Metabase a second time upsert values from `config.yml`, or no-op on duplicates? See discussion above
- [x] Should creating the first User **force** it to be an admin, even if you don't specify `is_superuser` (or explicitly specify `is_superuser: false`?

#### Follow-on PRs

- [ ] Document v1 of the config file spec, or come up with a plan for this 
- [ ] Add support for specifying Settings in a config file -- not sure if it should be this file or a different `settings.yml` file (mostly because it would be nice if those settings were available right away without having to wait until we run `metabase.config.file/initialize!`)
- [ ] Add a command to init from file without having to launch Metabase? Or otherwise make it easier to use in concert with SerDes

camsaul committed:
Initialize Metabase from a file

camsaul committed:
Code cleanup

camsaul committed:
Support optional template tags; more dox

camsaul commented:
Here's a loom that demos it. This is my first time trying to use Loom, and at this rate, almost certainly my last. My recording seems horribly broken. Have fun trying to watch it https://www.loom.com/share/e3cb28fcd48b4db28a46ed3bbb67c9cf

camsaul commented:
I tried doing another Loom with the Desktop app, the quality is every bit as garbage but in different ways. The text is a little clearer but the colors have horrible artifacts. https://www.loom.com/share/befaa35919d14b45a6fd8248121e9b6d

Watch them both and maybe you can figure out what's going on

bshepherdson reviewed (approved):
Looks great!

dpsutton commented:
I really love this idea. I only hesitate about how easy it is to make other users admin. Do you think that is a concern? Also, should we lock this down to just dev time for a release cycle to make sure it works? Or perhaps behind an env var at first? I'm feeling cautious but can't really point to a reason why though. So happy if you don't share these vague concerns.

metamben commented:
Maybe the entries could have a parameter deciding if an update should happen. I can imagine using this both strictly for initialization as well as for "enforcing" invariants (e.g., that certain components are present no matter what).

luizarakaki commented:
@dpsutton the main use case for this feature is to enable local development, so the user is able to create an admin anyway

metamben committed:
Remove data app permission code (#26016)

* Revert "Use "apps" collection namespace for app collections (#25963)"

This reverts commit a1894202ed1d5a2e80221e15e797f435ae2eab22.

* Revert "Implement EE/Pro app permissions (#25764)"

This reverts commit cf9b1bf50f2e025ef7a93f88d6a8e18cedfbba7c.

* Revert "Implement global app permissions for the "All Users" group (#25679)"

This reverts commit 63e950f256b115b8756ef631d075d681bb9b4114.

* Revert "Require data permissions for executing actions (#25784)"

This reverts commit 732bf6b14e76e9dbe5381346b9de243dceb2d4c9.

* Revert "Prevent modifying app permissions via the collection endpoint (#25684)"

This reverts commit 5d3f8279c0c1689ac958d778b9e5926e9bf0986b.

* Revert "Implement DB specific execution permissions (#25629)"

This reverts commit 141a13ce0bc89f77a02aeda6b0cfce2dd2f181b0.

* Revert "Introduce /execution/ permission (#25529)"

This reverts commit 57cbb921ba56d7e000df6af0cdada410b7591384.

* Keep the app permission revision table

snoe committed:
[Apps] Remove emitters (#25999)

* [Apps] Remove emitters

Emitters were created when we thought that we would be adding actions to
multiple places in metabase. Instead we've coalesced to actions being a
core feature of data-apps and are executed from within them. Attaching
actions to models, as a form of re-use rather than a form of execution,
means that actions can still be executed directly from a model, if we
want to allow that in the future.

* Fix lint errors

* Move perform-action! tests out of the api.

Since we want to keep the implicit bulk functionality but remove those
endpoints move the tests to the actions namespace.

* Fix linter

* Change driver test to use function instead of endpoint

camsaul committed:
Finish documentation for ns str

camsaul committed:
Force the first User to be an admin

camsaul committed:
Tests for DB creation

camsaul commented on src/metabase/config/file.clj:
```diff
@@ -0,0 +1,241 @@
+  ### TEMPLATES
+
+  After spec validation, the config map is walked and `{{template}}` forms are expanded. This uses the same code used
+  to parse template tags in SQL queries, i.e. [[metabase.driver.common.parameters.parse]], which
```
fixed

camsaul committed:
Add test to ensure that you can set User password with an {{env ...}} template

snoe committed:
Merge branch 'master' into init-from-file

ryan0x44 commented:
I've been following issue #2052 since mid-2018, and was delighted to see this merged this morning, thank you @camsaul ! 🥇
